test/test_cpp_api_parity.py

```diff
@@ -0,0 +1,456 @@
+    module_name = test_params_dict.get('module_name')
+    desc = test_params_dict.get('desc', None)
+    module_variant_name = module_name + (('_' + desc) if desc else '') + (('_' + device) if device != 'cpu' else '')
+    input_size = test_params_dict.get('input_size', None)
+    input_fn = test_params_dict.get('input_fn', None)
+    if input_size:
+        example_inputs = [torch.randn(input_size)]
+    elif input_fn:
+        example_inputs = list(input_fn())
+    else:
+        raise RuntimeError("Missing `input_size` or `input_fn` for {}".format(module_variant_name))
+    if device != 'cuda' or TEST_CUDA:
+        example_inputs = [x.to(device) for x in example_inputs]
+    return TorchNNTestParams(
+        module_name=module_name,
+        module_variant_name=module_variant_name,
+        python_constructor_args=test_params_dict.get('constructor_args'),
+        cpp_constructor_args=test_params_dict.get('cpp_constructor_args'),
+        example_inputs=example_inputs,
+        has_parity=test_params_dict.get('has_parity', True),
+        python_module_class=getattr(torch.nn, module_name),
+        cpp_sources=module_metadata.get('cpp_sources', ''),
+        device=device,
+    )
+
+
+def add_test(test_name, test_fn):
+    if hasattr(TestCppApiParity, test_name):
+        raise RuntimeError("Found two tests with the same name: " + test_name)
+    setattr(TestCppApiParity, test_name, test_fn)
+
+torch_nn_modules.module_metadata_map['SampleModule'] = sample_module.module_metadata
+
+torch_nn_test_params_map = {}
+
+torch_nn_module_names = set()
+
+for test_params_dict in sample_module.module_tests + common_nn.module_tests:
+    module_name = test_params_dict.get('module_name')
+    if module_name in torch_nn_has_parity:
+        torch_nn_module_names.add(module_name)
+        module_metadata = torch_nn_modules.module_metadata_map[module_name]
+        for device in ['cpu', 'cuda']:
+            test_params = _process_test_params(
+                test_params_dict=test_params_dict,
+                module_metadata=module_metadata,
+                device=device)
+            test_name = 'test_torch_nn_{}'.format(test_params.module_variant_name)
+            torch_nn_test_params_map[test_name] = test_params
+
+            def test_fn(self):
+                self._test_torch_nn_module_variant(test_params=torch_nn_test_params_map[self._testMethodName])
+
+            if device == 'cuda':
+                test_fn = unittest.skipIf(not TEST_CUDA, "CUDA unavailable")(test_fn)
+            add_test(test_name, test_fn)
+
+for module_name in sorted(list(torch_nn_module_names)):
+    ctor_args_test_name = 'test_torch_nn_{}_ctor_args'.format(module_name)
+
+    def ctor_args_test(self):
+        self._test_torch_nn_module_ctor_args(
+            module_name=self._testMethodName.replace('test_torch_nn_', '').replace('_ctor_args', ''))
+
+    add_test(ctor_args_test_name, ctor_args_test)
+
+
```

Comment:
Would be good to assert that there exists autogenerated tests for SampleModule, just in case someone messes up the test generation and the test trivially passes